# Introduction to SQL Using Python

In [1]:
import sqlite3
import pandas as pd

Download dataset here, https://www.kaggle.com/laudanum/footballdelphi.

In [2]:
# Connect to database
conn = sqlite3.connect('''database.sqlite''')

# Create cursor object
cur = conn.cursor()

In [ ]:
cur.execute('''Enter SQL query here;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
Unique_Teams_df.head()

For the first query, we will preview the contents of the table, __Unique_Teams__. To do this we can run the following query:

In [7]:
# View Unique_Teams dataframe

cur.execute('''SELECT * 
               FROM Unique_Teams;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
Unique_Teams_df

TeamName  Unique_Team_ID
0         Bayern Munich               1
1              Dortmund               2
2            Leverkusen               3
3            RB Leipzig               4
4            Schalke 04               5
5            M'gladbach               6
6             Wolfsburg               7
7               FC Koln               8
8            Hoffenheim               9
9                Hertha              10
10                Mainz              11
11              Hamburg              12
12        Werder Bremen              13
13        Ein Frankfurt              14
14             Augsburg              15
15             Freiburg              16
16            Stuttgart              17
17             Hannover              18
18           Ingolstadt              19
19            Darmstadt              20
20            Paderborn              21
21             Nurnberg              22
22         Braunschweig              23
23       Greuther Furth              24
24   Fortuna Dusseldorf              25
25       Kaiserslautern              26
26             St Pauli              27
27               Bochum              28
28            Karlsruhe              29
29            Bielefeld              30
..                  ...             ...
98           Man United              99
99                  QPR             100
100      Sheffield Weds             101
101             Swindon             102
102      Crystal Palace             103
103           Leicester             104
104       Nott'm Forest             105
105              Bolton             106
106       Middlesbrough             107
107               Derby             108
108          Sunderland             109
109            Barnsley             110
110            Charlton             111
111             Watford             112
112            Bradford             113
113              Fulham             114
114          Birmingham             115
115         Middlesboro             116
116           West Brom             117
117          Portsmouth             118
118              Wolves             119
119               Wigan             120
120             Reading             121
121                Hull             122
122               Stoke             123
123             Burnley             124
124           Blackpool             125
125             Swansea             126
126             Cardiff             127
127         Bournemouth             128

[128 rows x 2 columns]

The __SELECT__ statement tells the database what information we are trying to pull from the dataset. The asterisk, (*), tells the database that we want to select all the columns available in the table.
The __FROM__ statement tells the database that we want to select the table, __Unique_Teams__ to select data from.
The SQL query is ended by a semicolon, (;). This tell the database that you have ended your query similar to the way a period ends a sentence.

You can practice this on your own by trying to query all the contents of the __Matches__ table and comparing your query to the answer below.

In [8]:
# View Matches dataframe

cur.execute('''SELECT * 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID Div  Season        Date            HomeTeam        AwayTeam  \
0             1  D2    2009  2010-04-04          Oberhausen  Kaiserslautern   
1             2  D2    2009  2009-11-01         Munich 1860  Kaiserslautern   
2             3  D2    2009  2009-10-04       Frankfurt FSV  Kaiserslautern   
3             4  D2    2009  2010-02-21       Frankfurt FSV       Karlsruhe   
4             5  D2    2009  2009-12-06               Ahlen       Karlsruhe   
5             6  D2    2009  2010-04-03        Union Berlin       Karlsruhe   
6             7  D2    2009  2009-08-14           Paderborn       Karlsruhe   
7             8  D2    2009  2010-03-08           Bielefeld       Karlsruhe   
8             9  D2    2009  2009-09-26      Kaiserslautern       Karlsruhe   
9            10  D2    2009  2009-11-21       Hansa Rostock       Karlsruhe   
10           11  D2    2009  2009-12-19      Greuther Furth       Karlsruhe   
11           12  D2    2009  2010-04-16             Koblenz       Karlsruhe   
12           13  D2    2009  2010-03-21          Oberhausen       Karlsruhe   
13           14  D2    2009  2009-09-14             Cottbus       Karlsruhe   
14           15  D2    2009  2010-05-02            Duisburg       Karlsruhe   
15           16  D2    2009  2009-10-30  Fortuna Dusseldorf       Karlsruhe   
16           17  D2    2009  2010-01-15              Aachen       Karlsruhe   
17           18  D2    2009  2009-10-18            Augsburg       Karlsruhe   
18           19  D2    2009  2010-02-05            St Pauli       Karlsruhe   
19           20  D2    2009  2009-08-24         Munich 1860       Karlsruhe   
20           21  D1    2009  2010-02-06              Bochum      Leverkusen   
21           22  D1    2009  2009-11-22       Bayern Munich      Leverkusen   
22           23  D1    2009  2010-05-08          M'gladbach      Leverkusen   
23           24  D1    2009  2009-08-08               Mainz      Leverkusen   
24           25  D1    2009  2009-10-17             Hamburg      Leverkusen   
25           26  D1    2009  2010-04-17           Stuttgart      Leverkusen   
26           27  D1    2009  2010-03-20            Dortmund      Leverkusen   
27           28  D1    2009  2009-10-31          Schalke 04      Leverkusen   
28           29  D1    2009  2009-08-22            Freiburg      Leverkusen   
29           30  D1    2009  2010-02-21       Werder Bremen      Leverkusen   
...         ...  ..     ...         ...                 ...             ...   
24595     46745  E0    2016  2017-05-06             Swansea         Everton   
24596     46746  E0    2016  2017-05-07             Arsenal      Man United   
24597     46747  E0    2016  2017-05-07           Liverpool     Southampton   
24598     46748  E0    2016  2017-05-08             Chelsea   Middlesbrough   
24599     46749  E0    2016  2017-05-10         Southampton         Arsenal   
24600     46750  E0    2016  2017-05-12             Everton         Watford   
24601     46751  E0    2016  2017-05-12           West Brom         Chelsea   
24602     46752  E0    2016  2017-05-13         Bournemouth         Burnley   
24603     46753  E0    2016  2017-05-13            Man City       Leicester   
24604     46754  E0    2016  2017-05-13       Middlesbrough     Southampton   
24605     46755  E0    2016  2017-05-13               Stoke         Arsenal   
24606     46756  E0    2016  2017-05-13          Sunderland         Swansea   
24607     46757  E0    2016  2017-05-14      Crystal Palace            Hull   
24608     46758  E0    2016  2017-05-14           Tottenham      Man United   
24609     46759  E0    2016  2017-05-14            West Ham       Liverpool   
24610     46760  E0    2016  2017-05-15             Chelsea         Watford   
24611     46761  E0    2016  2017-05-16             Arsenal      Sunderland   
24612     46762  E0    2016  2017-05-16            Man City       West Brom   
24613     46763  E0    2016  2017-05-17         Southampt

There are 9 columns in the __Matches__ table. Perphaps we only want to select the column __Match_ID__. To do this we will just write __Match_ID__ in our select statement instead of the asterisk. The image below shows how this is done:

In [9]:
# View Match_ID from Matches

cur.execute('''SELECT Match_ID
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID
0             1
1             2
2             3
3             4
4             5
5             6
6             7
7             8
8             9
9            10
10           11
11           12
12           13
13           14
14           15
15           16
16           17
17           18
18           19
19           20
20           21
21           22
22           23
23           24
24           25
25           26
26           27
27           28
28           29
29           30
...         ...
24595     46745
24596     46746
24597     46747
24598     46748
24599     46749
24600     46750
24601     46751
24602     46752
24603     46753
24604     46754
24605     46755
24606     46756
24607     46757
24608     46758
24609     46759
24610     46760
24611     46761
24612     46762
24613     46763
24614     46764
24615     46765
24616     46766
24617     46767
24618     46768
24619     46769
24620     46770
24621     46771
24622     46772
24623     46773
24624     46774

[24625 rows x 1 columns]

Try selecting only the column __HomeTeam__ from the __Matches__ table and comparing your code to the query below:

In [10]:
# View HomeTeam from Matches

cur.execute('''SELECT HomeTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

HomeTeam
0              Oberhausen
1             Munich 1860
2           Frankfurt FSV
3           Frankfurt FSV
4                   Ahlen
5            Union Berlin
6               Paderborn
7               Bielefeld
8          Kaiserslautern
9           Hansa Rostock
10         Greuther Furth
11                Koblenz
12             Oberhausen
13                Cottbus
14               Duisburg
15     Fortuna Dusseldorf
16                 Aachen
17               Augsburg
18               St Pauli
19            Munich 1860
20                 Bochum
21          Bayern Munich
22             M'gladbach
23                  Mainz
24                Hamburg
25              Stuttgart
26               Dortmund
27             Schalke 04
28               Freiburg
29          Werder Bremen
...                   ...
24595             Swansea
24596             Arsenal
24597           Liverpool
24598             Chelsea
24599         Southampton
24600             Everton
24601           West Brom
24602         Bournemouth
24603            Man City
24604       Middlesbrough
24605               Stoke
24606          Sunderland
24607      Crystal Palace
24608           Tottenham
24609            West Ham
24610             Chelsea
24611             Arsenal
24612            Man City
24613         Southampton
24614           Leicester
24615             Arsenal
24616             Burnley
24617             Chelsea
24618                Hull
24619           Leicester
24620           Liverpool
24621          Man United
24622         Southampton
24623             Swansea
24624             Watford

[24625 rows x 1 columns]

You can select multiple columns by typing the name of each column in the select statement and seperating each column name by a comma, (,). You can see the names of both the Home Teams and Away Teams in each match below:

In [12]:
# View HomeTeam and AwayTeam from Matches

cur.execute('''SELECT HomeTeam,
                      AwayTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

HomeTeam        AwayTeam
0              Oberhausen  Kaiserslautern
1             Munich 1860  Kaiserslautern
2           Frankfurt FSV  Kaiserslautern
3           Frankfurt FSV       Karlsruhe
4                   Ahlen       Karlsruhe
5            Union Berlin       Karlsruhe
6               Paderborn       Karlsruhe
7               Bielefeld       Karlsruhe
8          Kaiserslautern       Karlsruhe
9           Hansa Rostock       Karlsruhe
10         Greuther Furth       Karlsruhe
11                Koblenz       Karlsruhe
12             Oberhausen       Karlsruhe
13                Cottbus       Karlsruhe
14               Duisburg       Karlsruhe
15     Fortuna Dusseldorf       Karlsruhe
16                 Aachen       Karlsruhe
17               Augsburg       Karlsruhe
18               St Pauli       Karlsruhe
19            Munich 1860       Karlsruhe
20                 Bochum      Leverkusen
21          Bayern Munich      Leverkusen
22             M'gladbach      Leverkusen
23                  Mainz      Leverkusen
24                Hamburg      Leverkusen
25              Stuttgart      Leverkusen
26               Dortmund      Leverkusen
27             Schalke 04      Leverkusen
28               Freiburg      Leverkusen
29          Werder Bremen      Leverkusen
...                   ...             ...
24595             Swansea         Everton
24596             Arsenal      Man United
24597           Liverpool     Southampton
24598             Chelsea   Middlesbrough
24599         Southampton         Arsenal
24600             Everton         Watford
24601           West Brom         Chelsea
24602         Bournemouth         Burnley
24603            Man City       Leicester
24604       Middlesbrough     Southampton
24605               Stoke         Arsenal
24606          Sunderland         Swansea
24607      Crystal Palace            Hull
24608           Tottenham      Man United
24609            West Ham       Liverpool
24610             Chelsea         Watford
24611             Arsenal      Sunderland
24612            Man City       West Brom
24613         Southampton      Man United
24614           Leicester       Tottenham
24615             Arsenal         Everton
24616             Burnley        West Ham
24617             Chelsea      Sunderland
24618                Hull       Tottenham
24619           Leicester     Bournemouth
24620           Liverpool   Middlesbrough
24621          Man United  Crystal Palace
24622         Southampton           Stoke
24623             Swansea       West Brom
24624             Watford        Man City

[24625 rows x 2 columns]

Practice selecting multiple columns by selecting __Match_ID__ and __Date__ from the __Matches__ table and compare your query to the code below:

In [14]:
# View Match_ID and Date from Matches

cur.execute('''SELECT Match_ID,
                      Date
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID        Date
0             1  2010-04-04
1             2  2009-11-01
2             3  2009-10-04
3             4  2010-02-21
4             5  2009-12-06
5             6  2010-04-03
6             7  2009-08-14
7             8  2010-03-08
8             9  2009-09-26
9            10  2009-11-21
10           11  2009-12-19
11           12  2010-04-16
12           13  2010-03-21
13           14  2009-09-14
14           15  2010-05-02
15           16  2009-10-30
16           17  2010-01-15
17           18  2009-10-18
18           19  2010-02-05
19           20  2009-08-24
20           21  2010-02-06
21           22  2009-11-22
22           23  2010-05-08
23           24  2009-08-08
24           25  2009-10-17
25           26  2010-04-17
26           27  2010-03-20
27           28  2009-10-31
28           29  2009-08-22
29           30  2010-02-21
...         ...         ...
24595     46745  2017-05-06
24596     46746  2017-05-07
24597     46747  2017-05-07
24598     46748  2017-05-08
24599     46749  2017-05-10
24600     46750  2017-05-12
24601     46751  2017-05-12
24602     46752  2017-05-13
24603     46753  2017-05-13
24604     46754  2017-05-13
24605     46755  2017-05-13
24606     46756  2017-05-13
24607     46757  2017-05-14
24608     46758  2017-05-14
24609     46759  2017-05-14
24610     46760  2017-05-15
24611     46761  2017-05-16
24612     46762  2017-05-16
24613     46763  2017-05-17
24614     46764  2017-05-18
24615     46765  2017-05-21
24616     46766  2017-05-21
24617     46767  2017-05-21
24618     46768  2017-05-21
24619     46769  2017-05-21
24620     46770  2017-05-21
24621     46771  2017-05-21
24622     46772  2017-05-21
24623     46773  2017-05-21
24624     46774  2017-05-21

[24625 rows x 2 columns]

There are 24,625 rows in the Matches table. This makes it difficult to find specific information. Perhaps we only wanted information from the 2015 Season. To do this, we can add a __WHERE__ statement to our SQL query. This allows us to select information that meets a certain condition. To query the information only pertaining to the 2015 season from the __Matches__ tables, we can use the query below:

In [15]:
# View HomeTeam and AwayTeam from Matches

cur.execute('''SELECT *
               FROM Matches
               WHERE Season = 2015;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID Div  Season        Date        HomeTeam        AwayTeam  FTHG  \
0        3540  D1    2015  2015-10-17   Werder Bremen   Bayern Munich     0   
1        3541  D1    2015  2016-03-05        Dortmund   Bayern Munich     0   
2        3542  D1    2015  2016-03-19         FC Koln   Bayern Munich     0   
3        3543  D1    2015  2016-01-22         Hamburg   Bayern Munich     1   
4        3544  D1    2015  2015-12-19        Hannover   Bayern Munich     0   
5        3545  D1    2015  2016-02-27       Wolfsburg   Bayern Munich     0   
6        3546  D1    2015  2016-04-09       Stuttgart   Bayern Munich     1   
7        3547  D1    2015  2015-09-26           Mainz   Bayern Munich     0   
8        3548  D1    2015  2016-02-06      Leverkusen   Bayern Munich     0   
9        3549  D1    2015  2016-02-14        Augsburg   Bayern Munich     1   
10       3550  D1    2015  2015-08-22      Hoffenheim   Bayern Munich     1   
11       3551  D1    2015  2016-05-07      Ingolstadt   Bayern Munich     1   
12       3552  D1    2015  2015-11-21      Schalke 04   Bayern Munich     1   
13       3553  D1    2015  2015-12-05      M'gladbach   Bayern Munich     3   
14       3554  D1    2015  2016-04-23          Hertha   Bayern Munich     0   
15       3555  D1    2015  2015-10-30   Ein Frankfurt   Bayern Munich     0   
16       3556  D1    2015  2015-09-19       Darmstadt   Bayern Munich     0   
17       3557  D1    2015  2016-04-10      Schalke 04        Dortmund     2   
18       3558  D1    2015  2016-04-23       Stuttgart        Dortmund     0   
19       3559  D1    2015  2015-10-04   Bayern Munich        Dortmund     5   
20       3560  D1    2015  2015-11-20         Hamburg        Dortmund     3   
21       3561  D1    2015  2015-12-05       Wolfsburg        Dortmund     1   
22       3562  D1    2015  2015-09-23      Hoffenheim        Dortmund     1   
23       3563  D1    2015  2016-03-02       Darmstadt        Dortmund     0   
24       3564  D1    2015  2016-01-23      M'gladbach        Dortmund     1   
25       3565  D1    2015  2015-08-23      Ingolstadt        Dortmund     0   
26       3566  D1    2015  2015-10-31   Werder Bremen        Dortmund     1   
27       3567  D1    2015  2016-03-20        Augsburg        Dortmund     1   
28       3568  D1    2015  2015-09-12        Hannover        Dortmund     2   
29       3569  D1    2015  2016-05-07   Ein Frankfurt        Dortmund     1   
..        ...  ..     ...         ...             ...             ...   ...   
962     46365  E0    2015  2016-04-30         Watford     Aston Villa     3   
963     46366  E0    2015  2016-04-30       West Brom        West Ham     0   
964     46367  E0    2015  2016-05-01      Man United       Leicester     1   
965     46368  E0    2015  2016-05-01     Southampton        Man City     4   
966     46369  E0    2015  2016-05-01         Swansea       Liverpool     3   
967     46370  E0    2015  2016-05-02         Chelsea       Tottenham     2   
968     46371  E0    2015  2016-05-07     Aston Villa       Newcastle     0   
969     46372  E0    2015  2016-05-07     Bournemouth       West Brom     1   
970     46373  E0    2015  2016-05-07  Crystal Palace           Stoke     2   
971     46374  E0    2015  2016-05-07       Leicester         Everton     3   
972     46375  E0    2015  2016-05-07         Norwich      Man United     0   
973     46376  E0    2015  2016-05-07      Sunderland         Chelsea     3   
974     46377  E0    2015  2016-05-07        West Ham         Swansea     1   
975     46378  E0    2015  2016-05-08       Liverpool         Watford     2   
976     46379  E0    2015  2016-05-08        Man City         Arsenal     2   
977     46380  E0    2015  2016-05-08       Tottenham     Southampton     1   
978     46381  E0    2015  2016-05-10        West Ham      Man United     3   
979     46382  E0    2015  2016-05-11       Liverpool         Chelsea     1   
980     46383  E0    2015  2016-05-11         Norwich  

Now we only have 992 rows and it allows us to see the dat from the 2015 season much easier than scrolling through the entire table as queired before.

In [4]:
# View Teams dataframe

cur.execute('''SELECT * FROM Teams;''')
Teams_df = pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df.head()

Season       TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0   2017  Bayern Munich        27         26                 15   
1   2017       Dortmund        33         25                 18   
2   2017     Leverkusen        31         24                 15   
3   2017     RB Leipzig        30         23                 15   
4   2017     Schalke 04        29         24                 17   

  OverallMarketValueHome AvgMarketValueHome StadiumCapacity  
0              597950000           22150000           75000  
1              416730000           12630000           81359  
2              222600000            7180000           30210  
3              180130000            6000000           42959  
4              179550000            6190000           62271

In [5]:
# View Teams_in_Matches dataframe

cur.execute('''SELECT * FROM Teams_in_Matches;''')
Teams_in_Matches_df = pd.DataFrame(cur.fetchall())
Teams_in_Matches_df.columns = [x[0] for x in cur.description]
Teams_in_Matches_df.head()

Match_ID  Unique_Team_ID
0         1              26
1         1              46
2         2              26
3         2              42
4         3              26

In [6]:
# View Unique_Teams dataframe

cur.execute('''SELECT * FROM Unique_Teams;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
Unique_Teams_df.head()

TeamName  Unique_Team_ID
0  Bayern Munich               1
1       Dortmund               2
2     Leverkusen               3
3     RB Leipzig               4
4     Schalke 04               5